### PR TITLE
Cargo.toml: Remove duplicate solana dependencies

### PR DESCRIPTION
**Pull request by ilya-bobyr**
#### Problem

A number of `solana-*` packages are included in the build tree twice.  Once at version `1.15.1` and the second time at version `1.16.0`.

See comment in `Cargo.toml` for an explanation of the issue, and a solution.

#### Summary of Changes
This change removes duplicate dependencies for older version of `solana-*` crates, and any of the dependencies that are no longer needed.

### CODE REVIEW

File: Cargo.toml
Comment (CriesofCarrots) [inline Cargo.toml:424]: I think it's not just "most likely", but always the case.

Comment (CriesofCarrots) [inline Cargo.toml:424]:
```suggestion
# what is available on crates.io.  As a consequence, if crates.io contains versions of these
# crates that are below the local version (which is most likely the case), we end up with
```

File: Cargo.toml
Comment (CriesofCarrots) [inline Cargo.toml:429]:
```suggestion
# There is a similar override in `programs/sbf/Cargo.toml`.  Please keep both comments and the
# overrides in sync.
```

File: programs/sbf/Cargo.toml
Comment (CriesofCarrots) [inline programs/sbf/Cargo.toml:188]:
```suggestion
# what is available on crates.io.  As a consequence, if crates.io contains versions of these
# crates that are below the local version (which is most likely the case), we end up with
```

File: programs/sbf/Cargo.toml
Comment (CriesofCarrots) [inline programs/sbf/Cargo.toml:193]:
```suggestion
# There is a similar override in `../../Cargo.toml`.  Please keep both comments and the
# overrides in sync.
```

**Review by CriesofCarrots — COMMENTED** [review]
I tested this approach with `cargo package` on v1.15.2, and it works great; shouldn't interfere with crate publishing. Wish we'd thought of this ages ago!
Just some comment typos to fix, otherwise lgtm. Would definitely be nice to get ✅ from your other reviewers too, though

File: Cargo.toml
Comment (ilya-bobyr) [reply Cargo.toml:424]: Fixed.

File: programs/sbf/Cargo.toml
Comment (ilya-bobyr) [reply programs/sbf/Cargo.toml:188]: Fixed.

File: programs/sbf/Cargo.toml
Comment (ilya-bobyr) [reply programs/sbf/Cargo.toml:193]: Fixed.

File: Cargo.toml
Comment (ilya-bobyr) [reply Cargo.toml:429]: Fixed.

**Comment by ilya-bobyr** [comment]
> I tested this approach with `cargo package` on v1.15.2, and it works great; shouldn't interfere with crate publishing.

Thank you!

> Wish we'd thought of this ages ago! Just some comment typos to fix, otherwise lgtm. Would definitely be nice to get :heavy_check_mark: from your other reviewers too, though

Thank you for proofreading.  Fixed both comments and rephrased them a bit.

**Comment by ilya-bobyr** [comment]
> Wish we'd thought of this ages ago!

Thinking a bit more about this, it seems that we want all the local packages to be specified in the `[patch]` section.  Making sure that all of our dependencies compile correctly with `master`, and always avoiding duplicate versions (which triggered the need for `FromOtherSolana` conversions).

I wonder why `cargo` is not behaving this way in the first place.
Maybe there are cases when such parsing could break version constrains, if done unconditionally.

**Comment by joncinque** [comment]
This definitely makes sense, but I'm worried we're going to end up needing to revert it if there's a breaking change in `solana-program` or `solana-zk-token-sdk`. The first one is good, since we don't want to make any breaking changes there, but the second is still in development. @samkim-crypto , are you planning any upcoming breaking changes there? If not, then we're good to go.

Also, how would this handle something like the upgrade to borsh 0.10 in #30975? That PR includes a a breaking upgrade to the borsh crate, so `spl-associated-token-account` would depend on `borsh` 0.9 *and* `solana-program` 1.16, so serialization will use two inconsistent crates. Thankfully, we're lucky that it doesn't try to serialize any of the `solana-program` types like `Pubkey`, so it might be ok. As a quick check, can you try cherry-picking that PR onto your branch to see if it still works?

If that's all good, then we can merge this, as long as we're aware that we still may have to revert it because of a breaking upgrade at some point.

File: bench-tps/src/spl_convert.rs
Comment (dmakarov) [inline bench-tps/src/spl_convert.rs:1]: Is removing this file, and subsequent deletions of `FromOtherSolana`, related to the changes in Cargo.toml, and the version of solana-program used by SPL? If not, maybe it would be better to make these two sets of changes as separate PRs, in case we need to revert Cargo.toml changes in the future?

Comment (dmakarov) [reply bench-tps/src/spl_convert.rs:1]: ok, read your later comment. never mind the noise.

**Review by dmakarov — APPROVED** [review]

**Comment by samkim-crypto** [comment]
There may be breaking changes in `solana-zk-token-sdk`, but these changes will always be accounted for on the SPL side first before the changes are merged on the monorepo side. So I think this should be fine for the zk-token-sdk at least.

**Comment by ilya-bobyr** [comment]
I have applied https://github.com/solana-labs/solana/pull/30975 on top of this change.
Here is a run with just the first commit: https://buildkite.com/solana-labs/solana/builds/95516
And here is a run with both the first and the second: https://buildkite.com/solana-labs/solana/builds/95522

I think, it is identical to what happens without this PR.

---

We had a discussion with @joncinque, and @samkim-crypto also joint at the end.
The conclusion is: this PR seems like a good thing to do.
While it might change the behavior in case of breaking changes, they are breaking changes in the current setup as well.

One idea, specifically for the `borsh` `0.10` migration case, is to provide derivations for both `0.9` and `0.10`.
Rather than moving completely to `0.10`.
This way the change is not breaking anymore.
Crates that want to use `borsh` `0.9` can still do so, until `0.9` derivations are removed in some future version.
But as this is additional work, Jon was thinking that it is probably fine to avoid it in this particular case.

---

I'll wait until Monday, to give everyone some time to respond.
But, it seems everyone is an agreement that this PR can be merged.

**Review by joncinque — APPROVED** [review]
Thanks for running the additional test, looks good to me then! And after our discussion, you're right that there are ways to resolve future breakage without reverting this.

**Comment by Skylerspencer** [comment]
But nine of the chances got or stopped solona from stilling my money through staking and fake votes that happened in the day you receive your rewards and when the program started it was only two validators me and 4,000 in capital you get it .and all I get for all these years is 20 dollars after my money generated millions for them

**Pull request merged**
